File: solutions/csharp/two-bucket/1/TwoBucket.cs
using FilledAmount = int;
using State = (int AmountStart, int AmountOther);
using WaterCapacity = int;

public enum Bucket
{
    One,
    Two
}

public readonly record struct TwoBucketResult(int Moves, Bucket GoalBucket, int OtherBucket);

public record TwoBucket(int CapacityOne, int CapacityTwo, Bucket Start)
{
    private readonly BucketInfo _start = BucketInfoFor(Start, CapacityOne, CapacityTwo);

    private readonly BucketInfo _other = BucketInfoFor(Other(Start), CapacityOne, CapacityTwo);

    public TwoBucketResult Measure(int goal)
    {
        EnsurePossible(goal);

        var start = InitialState();
        var queue = new Queue<(State State, int Moves)>();
        var seen = new HashSet<State>();

        queue.Enqueue((start, 1));
        seen.Add(start);

        while (queue.Count > 0)
        {
            var (state, moves) = queue.Dequeue();

            if (HasGoal(state, goal))
            {
                return ResultFor(state, moves, goal);
            }

            foreach (var next in NextStates(state).Where(IsAllowed))
            {
                if (seen.Add(next))
                {
                    queue.Enqueue((next, moves + 1));
                }
            }
        }

        throw new ArgumentException();
    }

    private State InitialState() => (_start.Capacity, 0);

    private TwoBucketResult ResultFor(State state, int moves, int goal) =>
        state.AmountStart == goal ? new(moves, _start.Id, state.AmountOther) : new(moves, _other.Id, state.AmountStart);

    private static bool HasGoal(State state, int goal) => state.AmountStart == goal || state.AmountOther == goal;

    private IEnumerable<State> NextStates(State state)
    {
        yield return (_start.Capacity, state.AmountOther);
        yield return (state.AmountStart, _other.Capacity);
        yield return (0, state.AmountOther);
        yield return (state.AmountStart, 0);
        yield return Pour(state.AmountStart, state.AmountOther, _other.Capacity);

        var (other, start) = Pour(state.AmountOther, state.AmountStart, _start.Capacity);
        yield return (start, other);
    }

    private static State Pour(FilledAmount source, FilledAmount target, WaterCapacity targetCapacity)
    {
        var amount = Math.Min(source, targetCapacity - target);
        return (source - amount, target + amount);
    }

    private bool IsAllowed(State state) => !_start.IsEmpty(state.AmountStart) || !_other.IsFull(state.AmountOther);

    private void EnsurePossible(int goal)
    {
        if (goal > Math.Max(CapacityOne, CapacityTwo) || goal % Gcd(CapacityOne, CapacityTwo) != 0)
        {
            throw new ArgumentException();
        }
    }

    private static int Gcd(int a, int b)
    {
        while (b != 0)
        {
            (a, b) = (b, a % b);
        }

        return a;
    }

    private static BucketInfo BucketInfoFor(Bucket bucket, WaterCapacity capacityOne, WaterCapacity capacityTwo) =>
        bucket == Bucket.One
            ? new(Bucket.One, capacityOne)
            : new(Bucket.Two, capacityTwo);

    private static Bucket Other(Bucket bucket) =>
        bucket == Bucket.One ? Bucket.Two : Bucket.One;

    private readonly record struct BucketInfo(Bucket Id, WaterCapacity Capacity)
    {
        public bool IsFull(FilledAmount amount) => amount == Capacity;

        public bool IsEmpty(FilledAmount amount) => amount == 0;
    }
}

File: solutions/python/two-bucket/1/two_bucket.py
"""Measure an exact amount of liquid using two buckets."""

from collections import deque
from math import gcd


def _next_states(state, capacities, start_bucket):
    first, second = state
    capacity_one, capacity_two = capacities

    poured_to_two = min(first, capacity_two - second)
    poured_to_one = min(second, capacity_one - first)
    states = (
        (capacity_one, second),
        (first, capacity_two),
        (0, second),
        (first, 0),
        (first - poured_to_two, second + poured_to_two),
        (first + poured_to_one, second - poured_to_one),
    )

    forbidden = (
        (0, capacity_two) if start_bucket == "one"
        else (capacity_one, 0)
    )
    return [candidate for candidate in states if candidate not in (state, forbidden)]


def measure(bucket_one, bucket_two, goal, start_bucket):
    """Return the minimum actions, goal bucket, and remaining amount."""
    if bucket_one <= 0 or bucket_two <= 0 or goal <= 0:
        raise ValueError("Bucket capacities and goal must be positive.")
    if start_bucket not in {"one", "two"}:
        raise ValueError("start_bucket must be 'one' or 'two'.")
    if goal > max(bucket_one, bucket_two) or goal % gcd(bucket_one, bucket_two):
        raise ValueError("Goal cannot be reached.")

    start = (bucket_one, 0) if start_bucket == "one" else (0, bucket_two)
    queue = deque([(start, 1)])
    visited = {start}

    while queue:
        state, actions = queue.popleft()
        first, second = state
        if goal in state:
            if first == goal:
                return actions, "one", second
            return actions, "two", first

        for next_state in _next_states(state, (bucket_one, bucket_two), start_bucket):
            if next_state not in visited:
                visited.add(next_state)
                queue.append((next_state, actions + 1))

    raise ValueError("Goal cannot be reached.")
